src/binding/vm.rs

```diff
@@ -20,3 +20,67 @@ pub fn require(name: &str) {
         vm::rb_require(name.as_ptr());
     }
 }
+
+
+pub fn thread_call_without_gvl<F, G>(func: F, unblock_func: Option<G>)
+    where F: FnMut(),
+          G: FnMut()
+{
+    unsafe {
+        vm::rb_thread_call_without_gvl(callbox as CallbackPtr,
+                                       Box::into_raw(Box::new(Box::new(func) as Box<FnMut()>)) as *const c_void,
+                                       callbox as CallbackPtr,
+                                       Box::into_raw(
+                                           Box::new(
+                                               unblock_func.map(|f| Box::new(f) as Box<FnMut()>).unwrap_or(Box::new(|| {}) as Box<FnMut()>)
+                                            )
+                                        ) as *const c_void)
+    }
+}
+
+pub fn thread_call_without_gvl2<F, G>(func: F, unblock_func: Option<G>)
+    where F: FnMut(),
+          G: FnMut()
+{
+    unsafe {
+        vm::rb_thread_call_without_gvl2(callbox as CallbackPtr,
+                                       Box::into_raw(Box::new(Box::new(func) as Box<FnMut()>)) as *const c_void,
+                                       callbox as CallbackPtr,
+                                       Box::into_raw(
+                                           Box::new(
+                                               unblock_func.map(|f| Box::new(f) as Box<FnMut()>).unwrap_or(Box::new(|| {}) as Box<FnMut()>)
+                                            )
+                                        ) as *const c_void)
+    }
+}
+
+pub fn thread_call_with_gvl<F>(func: F)
+    where F: FnMut()
+{
+    unsafe {
+        vm::rb_thread_call_with_gvl(callbox as CallbackPtr,
+                                    Box::into_raw(Box::new(Box::new(func) as Box<FnMut()>)) as *const c_void);
+    }
+}
+
+extern "C" fn callbox(boxptr: *mut c_void) {
+    let mut fnbox: Box<Box<FnMut()>> = unsafe { Box::from_raw(boxptr as *mut Box<FnMut()>) };
+    fnbox();
+}
+
+
+pub fn protect<F>(func: F) -> Result<(), ()>
```

Comment:
I think the return type of this function should be `Result<Value, i32>` to represent the [actual result of `rb_protect`](https://github.com/ruby/ruby/blob/6a69ab937c7e6584d311cd444615cc1cd5b5499b/doc/extension.rdoc#control-structure).

 - `Ok(value)` in case of success
 - `Err(state)` in case of failure